### PR TITLE
Add new rule FavourStaticEmptyFields

### PR DESCRIPTION
**Pull request by janus**
It favours use of static empty fields.

### CODE REVIEW

File: src/FSharpLint.Core/Text.resx [hunk 1]
```diff
@@ -333,4 +333,7 @@
   <data name="RulesFavourConsistentThis" xml:space="preserve">
     <value>Prefer using '{0}' consistently.</value>
   </data>
+  <data name="RulesFavourStaticEmptyFieldsError" xml:space="preserve">
+    <value>Use a static empty field such as String.Empty or List.Empty</value>
```
Comment (knocte) [inline src/FSharpLint.Core/Text.resx:337]: Please be more specific here by using `{0}` inside the string, and injecting "String.Empty" or "List.Empty" depending on what violation happens.

Comment (knocte) [reply src/FSharpLint.Core/Text.resx:337]: (as an example of how this is done, you can look at the rule "FailwithBadUsage")

File: tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        this.Parse "System.Console.WriteLine List.Empty.Length"
+
+        Assert.IsTrue this.NoErrorsExist
+
+    [<Test>]
+    member this.FavourStaticEmptyFieldsShouldNotProduceError_7() =
+        this.Parse """
+match foo with
+| List.Empty -> true
+| head::_ -> false"""
```
Comment (knocte) [inline tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs:90]: this F# snippet doesn't compile, therefore this test should be deleted

File: tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        Assert.IsTrue this.ErrorsExist
+
+    [<Test>]
+    member this.FavourStaticEmptyFieldsShouldProduceError_6() =
+        this.Parse """
+match foo with
+| [] -> true
+| head::_ -> false"""
+
+        Assert.IsTrue this.ErrorsExist
```
Comment (knocte) [inline tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs:53]: given that the F# snippet in the test FavourStaticEmptyFieldsShouldNotProduceError_7 doesn't compile, then this Assert should be the opposite: NoErrorsExist

File: src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+open System.Collections.Generic
+
+let private getStaticEmptyErrorMessage  (range:FSharp.Compiler.Text.Range) (idText:string) =
+    let formatError errorName =
+        String.Format(Resources.GetString errorName, idText)
+
+    "RulesFavourStaticEmptyFieldsError" |> formatError |> Array.singleton
+
+let private generateError (range:FSharp.Compiler.Text.Range) (idText:string) =
+    if idText = "" || idText = "[]" then
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs:19]: please also include `[| |]`, which should be caught by the rule and suggest Array.empty instead

File: docs/content/how-tos/rules/FL0075.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+hide_menu: true
+---
+
+# FavourStaticEmptyFields (FL0075)
+
+*Introduced in `0.21.1`*
+
+## Cause
+
+Rule to detect use of non static empty fields.
```
Comment (knocte) [inline docs/content/how-tos/rules/FL0075.md:13]: The fact that this rule recommends static empty fields doesn't mean that the flagged cases are non-static. The world is not black or white.

File: src/FSharpLint.Core/Text.resx (hunk 1 above)
Comment (janus) [reply src/FSharpLint.Core/Text.resx:337]: Done

File: tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs (hunk 3 above)
Comment (janus) [reply tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs:56]: Done

File: docs/content/how-tos/rules/FL0075.md (hunk 5 above)
Comment (janus) [reply docs/content/how-tos/rules/FL0075.md]: Done

File: tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    member this.FavourStaticEmptyFieldsShouldNotProduceError_7() =
+        this.Parse "System.Console.WriteLine Array.empty.Length"
+
+        Assert.IsTrue this.NoErrorsExist
+
+    [<Test>]
+    member this.FavourStaticEmptyFieldsShouldNotProduceError_8() =
+        this.Parse """
+match foo with
+| [] -> true| head::_ -> false"""
```
Comment (knocte) [inline tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs:104]: what syntax style is this? please put an EOL after `true`

File: docs/content/how-tos/rule-configuration.md [hunk 7]
```diff
@@ -114,4 +114,5 @@ The following rules can be specified for linting.
 - [CyclomaticComplexity (FL0071)](rules/FL0071.html)
 - [FailwithBadUsage (FL0072)](rules/FL0072.html)
 - [FavourReRaise (FL0073)](rules/FL0073.html)
-- [FavourConsistentThis (FL0074)](rules/FL0074.html)
```
Comment (knocte) [inline docs/content/how-tos/rule-configuration.md:117]: about the commit message: the character * is used when writing a list of things, but here there is only one line, not many, so why do you use the * character? there's no need for it

**Comment by knocte** [comment]
You need to rebase this PR.

**Comment by knocte** [comment]
@janus can you remove the "WIP" commit from your `janus:janus/favourStaticEmptyFields` branch (as it is not needed) and after that, rebase your branch (rebase, not merge)? Thanks

**Comment by janus** [comment]
> @janus can you remove the "WIP" commit from your `janus:janus/favourStaticEmptyFields` branch (as it is not needed) and after that, rebase your branch (rebase, not merge)? Thanks

Thanks. Done.

**Comment by knocte** [comment]
> Thanks. Done.

Not true, I still see conflicts in this PR.

**Comment by knocte** [comment]
> Not true, I still see conflicts in this PR.

And I still see the WIP commit, so not sure what you have done.

**Comment by janus** [comment]
> > Not true, I still see conflicts in this PR.
> 
> And I still see the WIP commit, so not sure what you have done.

I did. But it is here, https://github.com/janus/FSharpLint/commits/favourStaticEmptyFields

**Comment by knocte** [comment]
> But it is here, https://github.com/janus/FSharpLint/commits/favourStaticEmptyFields

And why do you push it there? This PR is not targetting that branch. I told you in which branch you should do it: in the branch for this PR.

**Comment by janus** [comment]
> > But it is here, https://github.com/janus/FSharpLint/commits/favourStaticEmptyFields
> 
> And why do you push it there? This PR is not targetting that branch. I told you in which branch you should do it: in the branch for this PR.

It happened that I have deleted the branch from my local repo. I have to pick it up from remote which detached it. That's why it is not linked to the PR.

**Comment by knocte** [comment]
Why do you delete branches? Especially branches which are targeted by PRs that are not merged yet?

**Comment by janus** [comment]
> Why do you delete branches? Especially branches which are targeted by PRs that are not merged yet?

Some weeks ago I noticed that I have about 30 branches in my local repo. It makes it had to focus on current branches. So, I decided to remove ones that are months old. I erred for not first checking if that are already merged before deleting.

**Comment by knocte** [comment]
> I have to pick it up from remote which detached it

I still don't see it.

**Comment by janus** [comment]
> > I have to pick it up from remote which detached it
> 
> I still don't see it.

It is there now.

**Comment by knocte** [comment]
> It is there now.

And I still see conflicts in the PR, so it is not rebased yet.

**Comment by janus** [comment]
> > It is there now.
> 
> And I still see conflicts in the PR, so it is not rebased yet.

It is re-based now.

File: docs/content/how-tos/rules/FL0076.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+hide_menu: true
+---
+
+# FavourStaticEmptyFields (FL0076)
+
+*Introduced in `0.21.1`*
+
+## Cause
+
+Rule favours static empty fields such as String.Empty, List.Empty and also Array.Empty.
```
Comment (knocte) [inline docs/content/how-tos/rules/FL0076.md:13]: Please replace this sentence with `Use of immediate string "", empty list [] or empty array [||]`.

File: docs/content/how-tos/rules/FL0076.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+## Rationale
+
+Using static empty fields aids readibility.
+
+## How To Fix
+
+Use a static empty field such as String.Empty or List.Empty and Array.Empty.
```
Comment (knocte) [inline docs/content/how-tos/rules/FL0076.md:21]: `a or b and c`? I guess you mean `a, b or c` or `a, b and c`, but not a mix.

File: src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs (hunk 4 above)
Comment (knocte) [reply src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs:38]: @janus you didn't address this yet

Comment (janus) [reply src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs:38]: It is already acted upon. Sorry that I didn't flag resolved. Check test and code base if in doubt.

File: docs/content/how-tos/rules/FL0076.md (hunk 8 above)
Comment (janus) [reply docs/content/how-tos/rules/FL0076.md]: Already resolved.

File: docs/content/how-tos/rules/FL0076.md (hunk 9 above)
Comment (janus) [reply docs/content/how-tos/rules/FL0076.md]: Done. Thanks.

File: tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs (hunk 6 above)
Comment (janus) [reply tests/FSharpLint.Core.Tests/Rules/Conventions/FavourStaticEmptyFields.fs]: Done. Thanks.

File: docs/content/how-tos/rules/FL0076.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+hide_menu: true
+---
+
+# FavourStaticEmptyFields (FL0076)
+
+*Introduced in `0.21.1`*
+
+## Cause
+
+Use of immediate string "", empty list [] and empty array [||]
```
Comment (knocte) [inline docs/content/how-tos/rules/FL0076.md:13]: I said `or`, not `and`. Also, this is a paragraph so it should end with a dot.

Comment (janus) [reply docs/content/how-tos/rules/FL0076.md]: Done. Sorry.

File: src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    | _ -> Array.empty
+
+
+let rule =
+    { Name = "FavourStaticEmptyFields"
+      Identifier = Identifiers.FavourStaticEmptyFields
+      RuleConfig =
+        { AstNodeRuleConfig.Runner = runner
+          Cleanup = ignore } }
+    |> AstNodeRule
```
Comment (knocte) [inline src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs:98]: nit: please end file with \n\n

File: docs/content/how-tos/rule-configuration.md [hunk 12]
```diff
@@ -115,4 +115,5 @@ The following rules can be specified for linting.
 - [FailwithBadUsage (FL0072)](rules/FL0072.html)
 - [FavourReRaise (FL0073)](rules/FL0073.html)
 - [FavourConsistentThis (FL0074)](rules/FL0074.html)
-- [AvoidTooShortNames (FL0075)](rules/FL0075.html)
\ No newline at end of file
+- [AvoidTooShortNames (FL0075)](rules/FL0075.html)
+- [FavourStaticEmptyFields (FL0075)](rules/FL0076.html)
```
Comment (knocte) [inline docs/content/how-tos/rule-configuration.md:119]: @janus should be 75, not 76, sigh

Comment (janus) [reply docs/content/how-tos/rule-configuration.md]: Done. Thanks.

File: src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs [hunk 13]
```diff
@@ -0,0 +1,98 @@
+module FSharpLint.Rules.FavourStaticEmptyFields
+
+open FSharpLint.Framework
+open FSharpLint.Framework.Suggestion
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.Text
+open FSharpLint.Framework.Ast
+open FSharpLint.Framework.Rules
+open System
+open System.Collections.Generic
```
[the remark this answers is not in the document]
Comment (janus) [reply src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs]: Removed. Thanks.

File: src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs (hunk 11 above)
Comment (janus) [reply src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs]: Applied. Thanks.

File: src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs (hunk 4 above)
Comment (knocte) [reply src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs:36]: how about `[ ]`? (with a space in the middle); I don't see a test for that

Comment (janus) [reply src/FSharpLint.Core/Rules/Conventions/FavourStaticEmptyFields.fs:36]: This state would be seen ass [](which is already there), however let me add it in the test.

**Pull request merged**